File: src/main/java/net/draconia/askaround/domain/State.java
package net.draconia.askaround.domain;

import java.io.Serializable;

public class State implements Serializable
{
	private static final long serialVersionUID = 6645877398852435359L;
	
	private Country mObjCountry;
	private Integer miId;
	private String msAbbreviation, msName;
	
	public State()
	{ }
	
	public State(final String sName)
	{
		setName(sName);
	}
	
	public State(final Integer iId, final String sName)
	{
		setId(iId);
		
		setName(sName);
	}
	
	public State(final String sName, final String sAbbreviation)
	{
		this(sName);
		
		setAbbreviation(sAbbreviation);
	}
	
	public State(final String sName, final Country objCountry)
	{
		this(sName);
		
		setCountry(objCountry);
	}
	
	public State(final Integer iId, final String sName, final String sAbbreviation)
	{
		this(iId, sName);
		
		setAbbreviation(sAbbreviation);
	}
	
	public State(final String sName, final String sAbbreviation, final Country objCountry)
	{
		this(sName, sAbbreviation);
		
		setCountry(objCountry);
	}
	
	public State(final Integer iId, final String sName, final Country objCountry)
	{
		this(iId, sName);
		
		setCountry(objCountry);
	}
	
	public State(final Integer iId, final String sName, final String sAbbreviation, final Country objCountry)
	{
		this(iId, sName, sAbbreviation);
		
		setCountry(objCountry);
	}
	
	public String getAbbreviation()
	{
		if(msAbbreviation == null)
			msAbbreviation = "";
		
		return(msAbbreviation);
	}
	
	public Country getCountry()
	{
		if(mObjCountry == null)
			mObjCountry = new Country();
		
		return(mObjCountry);
	}
	
	public int getId()
	{
		if(miId == null)
			miId = 0;
		
		return(miId);
	}
	
	public String getName()
	{
		if(msName == null)
			msName = "";
		
		return(msName);
	}
	
	public void setAbbreviation(final String sAbbreviation)
	{
		if(sAbbreviation == null)
			msAbbreviation = "";
		else
			msAbbreviation = sAbbreviation;
	}
	
	public void setCountry(final Country objCountry)
	{
		if(objCountry == null)
			mObjCountry = new Country();
		else
			mObjCountry = objCountry;
	}
	
	public void setId(final Integer iId)
	{
		if(iId == null)
			miId = 0;
		else
			miId = iId;
	}
	
	public void setName(final String sName)
	{
		if(sName == null)
			msName = "";
		else
			msName = sName;
	}
}

File: src/main/webapp/scripts/models/state.js
function State()
{
	this.msAbbreviation					=	undefined;
	this.mbChanged						=	undefined;
	this.mObjCountry					=	undefined;
	this.miId							=	undefined;
	this.msName							=	undefined;
	this.mArrObjObservers				=	[];
}

State.prototype.addObserver				=	function(objObserver)
{
	if(objObserver != undefined)
		{
		if(this.mArrObjObservers == undefined)
			this.mArrObjObservers 		=	[];
		
		this.mArrObjObservers.push(objObserver);
		}
	
	return(this);
};

State.copyFromJSON						=	function(objJSON, objState)
{
	if(objState == undefined)
		objState						=	new State();
	
	for(prop in objJSON)
		switch(prop)
			{
			case "country":
				objState.setCountry(Country.copyFromJSON(eval("objJSON." + prop), objState.getCountry()));
				break;
			default:
				eval("objState.set" + prop.substring(0, 1).toUpperCase() + prop.substring(1) + "(objJSON." + prop + ");");
				break;
			}
	
	return(objState);
};

State.prototype.getAbbreviation			=	function()
{
	if(this.msAbbreviation == undefined)
		this.msAbbreviation				=	"";
	
	return(this.msAbbreviation);
};

State.prototype.getCountry				=	function()
{
	if(this.mObjCountry == undefined)
		this.mObjCountry				=	new Country();
	
	return(this.mObjCountry);
};

State.prototype.getId					=	function()
{
	if(this.miId == undefined)
		this.miId						=	0;
	
	return(this.miId);
};

State.prototype.getName					=	function()
{
	if(this.msName == undefined)
		this.msName						=	"";
	
	return(this.msName);
};

State.prototype.hasChanged				=	function()
{
	if(this.mbChanged == undefined)
		this.mbChanged = false;
	
	return(this.mbChanged);
};

State.prototype.notifyObservers			=	function(objArgument)
{
	this.mbChanged = true;
	
	for(var iLength = this.mArrObjObservers.length, iLoop = 0; iLoop < iLength; iLoop++)
		this.mArrObjObservers[iLoop].call(this, objArgument);
	
	this.mbChanged = false;
};

State.prototype.packageToJSON			=	function()
{
	return(	{	id:				this.getId()
			,	name:			this.getName()
			,	abbreviation:	this.getAbbreviation()
			,	country:		this.getCountry().packageToJSON()
			});
};

State.prototype.setAbbreviation			=	function(sAbbreviation)
{
	if(sAbbreviation == undefined)
		this.msAbbreviation				=	"";
	else
		this.msAbbreviation				=	sAbbreviation;
	
	this.notifyObservers();
	
	return(this);
};

State.prototype.setCountry				=	function(objCountry)
{
	if(objCountry == undefined)
		this.mObjCountry				=	new Country();
	else
		this.mObjCountry				=	objCountry;
	
	this.notifyObservers();
	
	return(this);
};

State.prototype.setId					=	function(iId)
{
	if(iId == undefined)
		this.miId						=	0;
	else
		this.miId						=	iId;
	
	this.notifyObservers();
	
	return(this);
};

State.prototype.setName					=	function(sName)
{
	if(sName == undefined)
		this.msName						=	"";
	else
		this.msName						=	sName;
	
	this.notifyObservers();
	
	return(this);
};